src/main/kotlin/com/exactpro/th2/codec/EventProcessor.kt

```diff
@@ -0,0 +1,146 @@
+) {
+    fun onEvent(message: String, messagesIds: List<MessageID> = emptyList()) = codecEventID.onEvent(message, messagesIds)
+    fun onErrorEvent(message: String, messagesIds: List<MessageID> = emptyList(), cause: Throwable? = null) = codecEventID.onErrorEvent(message, messagesIds, cause)
+
+    private fun EventID.onEvent(
+        message: String,
+        messagesIds: List<MessageID> = emptyList(),
+        body: List<String> = emptyList(),
+    ) : EventID {
+        LOGGER.warn { "$message. Messages: ${messagesIds.joinToString(transform = MessageID::logId)}" }
+        return this.publishEvent(message, messagesIds, body = body).id
+    }
+
+    private fun EventID.onErrorEvent(
+        message: String,
+        messagesIds: List<MessageID> = emptyList(),
+        cause: Throwable? = null,
+        additionalBody: List<String> = emptyList()
+    ): EventID {
+        LOGGER.error(cause) { "$message. Messages: ${messagesIds.joinToString(transform = MessageID::logId)}" }
+        return publishEvent(message, messagesIds, Event.Status.FAILED, cause, additionalBody).id
+    }
+
+    fun onEachEvent(
+        parentEventIDs: Sequence<EventID>,
+        message: String,
+        messagesIds: List<MessageID> = emptyList(),
+        body: List<String> = emptyList()
+    ) {
+        val warnEvent = codecEventID.onEvent(message, messagesIds, body)
+        parentEventIDs.forEach {
+            it.addReferenceTo(warnEvent, message, Event.Status.PASSED)
+        }
+    }
+
+    fun onEachErrorEvent(
+        parentEventIDs: Sequence<EventID>,
+        message: String,
+        messagesIds: List<MessageID> = emptyList(),
+        cause: Throwable? = null,
+        additionalBody: List<String> = emptyList(),
+    ): EventID {
+        val errorEventId = codecEventID.onErrorEvent(message, messagesIds, cause, additionalBody)
+        parentEventIDs.forEach { it.addReferenceTo(errorEventId, message, Event.Status.FAILED) }
+        return errorEventId
+    }
+
+    fun onEachWarning(
+        parentEventIDs: Sequence<EventID>,
+        context: ReportingContext,
+        action: String,
+        additionalBody: () -> List<String> = ::emptyList,
+        messagesIds: () -> List<MessageID> = ::emptyList
+    ) = context.warnings.let { warnings ->
+        if (warnings.isNotEmpty()) {
+            val messages = messagesIds()
+            val body = additionalBody()
+            warnings.forEach { warning ->
+                onEachEvent(parentEventIDs, "[WARNING] During $action: $warning", messages, body)
+            }
+        }
+    }
+
+    private fun EventID.addReferenceTo(eventId: EventID, name: String, status: Event.Status): EventID = Event.start()
+        .endTimestamp()
+        .name(name)
+        .status(status)
+        .type(if (status != Event.Status.PASSED) "Error" else "Warn")
+        .bodyData(EventUtils.createMessageBean("This event contains reference to the codec event"))
+        .bodyData(ReferenceToEvent(eventId.logId))
```

Comment:
`logId` is for logging. I am not sure if we can use it like this (it always can be changed). I believe we must to it by ourselves here